docs/config.md
# Configuring an OCSP Responder
The following YAML format configuration includes all the configurations for `dycosp`.
## List of configuration options
```yaml
version: 0.1
strict: false
expiration: "ignore"
log:
  level: "info"
  format: "json"
responder:
  ca: "sub-ca"
  responder_certificate: "dyocsp/testdata/sub-ocsp-rsa.crt"
  responder_key: "dyocsp/testdata/sub-ocsp-rsa-pkcs8.key"
  issuer_certificate: "dyocsp/testdata/sub-ca-rsa.crt"
cache:
  interval: 60
  delay: 5
db:
  dynamodb:
    region: "us-west-2"
    table_name: "test_ca_db"
    ca_gsi: "ca_gsi"
    endpoint: ""
    retry_max_attempts: 0
    timeout: 60
  file:
    file: "testdata/filedb"
http:
  addr: ""
  port: 80
  read_timeout: 30
  write_timeout: 0
  read_header_timeout: 10
  max_header_bytes: 1048576 1M
  max_request_bytes: 256
  cache_control_max_age: 60 (same as cache.interval)
```
## version
```yaml
version: 0.1
```
The `version` option is a __required__ field that specifies the configuration version.

## strict
```yaml
strict: false
```
The `strict` option is an __optional (default: false)__ field that enables strict mode.
When strict mode is enabled, if the DB client encounters an error, the `dyocsp` server should stop with a panic.

## expiration
```yaml
expiration: "ignore"
```
Configures the behavior when the "Expiration Date" of a certificate is
exceeded. If set to "ignore", the "Expiration Date" will be ignored, and a response will
be generated. This is the default setting.
To output a warning at log level Warn, set it to "warn". If set to "invalid",
no response will be generated.

## log
```yaml
log:
  level: "info"
  format: "json"
```
The `log` section configures the log behavior.

|Parameter|Required|Default|Description|
| ----------- | ----------- | ----------- | ----------- |
|level|no|`info`|Log level selected in `error`, `warn`, `info`, `debug`.|
|format|no|`json`|Log format selected in either `json` or `pretty`. The `pretty` format is a human-readable (read on terminal) format.|

## responder
```yaml
responder:
  ca: "sub-ca"
  responder_certificate: "dyocsp/testdata/sub-ocsp-rsa.crt"
  responder_key: "dyocsp/testdata/sub-ocsp-rsa-pkcs8.key"
  issuer_certificate: "dyocsp/testdata/sub-ca-rsa.crt"
```
The `responder` section of the configuration file is used to configure the self
 certificate, key, and issuer for the OCSP responder.
Private key `responder_key` can also be set from the `DYOCSP_PRIVATE_KEY` environment  variable.
However, this configuration is mutually exclusive, and an error occurs if the configuration is duplicated.

|Parameter|Required|Default|Description|
| ----------- | ----------- | ----------- | ----------- |
|ca|yes||`ca` can be used as an index key for tables or data structures in a database.|
|responder_certificate|yes||The path to the responder's certificate.|
|responder_key|yes||The path to the responder's private key. |
|issuer_certificate|yes||The path to the certificate issuer's certificate. |

## cache
```yaml
cache:
  interval: 60
  delay: 5
```
`cache` section configures the life cycle of pre-generated OCSP response caches.
Please refer to the [cache lifecycle](cache_lifecycle.md) document for detailed information about cache.

|Parameter|Required|Default|Description|
| ----------- | ----------- | ----------- | ----------- |
|interval|no|60 (sec)|`interval` configures the duration between `nextUpdate` and `nextUpdate`. The units are in seconds.|
|delay|no|5 (sec)|`delay` configures the duration of delay processing before reaching `nextUpdate`. The units are in seconds.|

## db
```yaml
db:
  dynamodb:
    region: "us-west-2"
    table_name: "test_ca_db"
    ca_gsi: "ca_gsi"
    endpoint: ""
    retry_max_attempts: 0
    timeout: 60
  file:
    file: "testdata/filedb"
```
`db` section configures the type of database and the configuration parameters for the selected database.
Type of database is exclusive, and if the type is duplicated, an error occurs.

### dynamodb
Please refer to the [dynamodb](dynamodb.md) documentation for details about this database type.

|Parameter|Required|Default|Description|
| ----------- | ----------- | ----------- | ----------- |
|region|yes||The DynamoDB Region.|
|table_name|yes||The Name of the table that has revocation information and `dyocsp` will access.|
|ca_gsi|yes||The Name of the required global secondary index.|
|endpoint|no||The Endpoint URL of DynamoDB should be set when using the local DynamoDB server.|
|retry_max_attempts|no|0|`retry_max_attempts` specifies the maximum number attempts an API client will call an operation that fails with a retryable error. A value of 0 is ignored.|
|timeout|no|60|`timeout` parameter specifies the timeout for the API client request.|

### file
Prease refer [file db](fileasdb.md) documentation for details about this database type.

|Parameter|Required|Default|Description|
| ----------- | ----------- | ----------- | ----------- |
|file|yes||The path to file DB.|

## http
```yaml
http:
  addr: ""
  port: 80
  read_timeout: 30
  write_timeout: 0
  read_header_timeout: 10
  max_header_bytes: 1048576 1M
  max_request_bytes: 256
  cache_control_max_age: 60
```
`http` section configures the behavior of the HTTP server.
|Parameter|Required|Default|Description|
| ----------- | ----------- | ----------- | ----------- |
|addr|no||The address for the server to listen on.|
|port|no|80|The port number for the server to listen on.|
|read_timeout|no|30|`read_timeout` is the maximum duration for reading the entire request, including the body. A zero or negative value means there will be no timeout.|
|write_timeout|no|0|`write_timeout` is the maximum duration before timing out writes of the response. A zero or negative value means there will be no timeout.|
|read_header_timeout|no|10|`read_header_timeout` is the amount of time allowed to read request headers. If `read_header_timeout` is zero, the value of ReadTimeout is used.|
|max_header_bytes|no|1048576 (1M)|`max_header_bytes` controls the maximum number of bytes the server will read parsing the request header's keys and values, including the request line. It does not limit the size of the request body.|
|max_request_bytes|no|256|`max_request_bytes` defines the maximum size of a request in bytes. Since the content of an OCSP request has a fixed form, the default value is as small as 256 bytes.|
|cache_control_max_age|no|60|`cache_control_max_age` defines the maximum age, in seconds, for a cached response as specified in the Cache-Control max-age directive. If the duration until the nextUpdate of a cached response exceeds MaxAge, the handler sets the response's Cache-Control max-age directive to that duration.|


pkg/config/config.go
package config

import (
	"fmt"
	"regexp"

	"github.com/rs/zerolog"
)

// The DyOCSPConfig struct contains configuration members for creating instances
// of DyOCSP. Please refer to the documentation for detailed information about each
// configuration.
type DyOCSPConfig struct {
	// From ConfigYAML
	Version                  string
	Strict                   bool
	Expiration               string
	LogLevel                 string
	LogFormat                string
	CA                       string
	Certificate              string
	Key                      string
	Issuer                   string
	Interval                 int
	Delay                    int
	DynamoDBRegion           string
	DynamoDBTableName        string
	DynamoDBCAGsi            string
	DynamoDBEndpoint         string
	DynamoDBRetryMaxAttempts int
	DynamoDBTimeout          int
	FileDBFile               string
	Port                     string
	Domain                   string
	ReadTimeout              int
	WriteTimeout             int
	ReadHeaderTimeout        int
	MaxHeaderBytes           int
	MaxRequestBytes          int
	CacheControlMaxAge       int
	// From this struct
	ZerologLevel  zerolog.Level
	ZerologFormat LogFormat
	DBType        CADBType
}

// The ConfigYAML is a configuration file in YAML format.
// To indicate a non-specified status, the member of type int should be a pointer.
// This struct instance verifies the instance's own members and creates a DyOCSPConfig
// object based on the instance's attributes.
type ConfigYAML struct {
	Version    string `yaml:"version"`
	Strict     bool   `yaml:"strict"`
	Expiration string `yaml:"expiration"`
	Log        struct {
		Level  string `yaml:"level"`
		Format string `yaml:"format"`
	} `yaml:"log"`
	Responder struct {
		CA          string `yaml:"ca"`
		Certificate string `yaml:"responder_certificate"`
		Key         string `yaml:"responder_key"`
		Issuer      string `yaml:"issuer_certificate"`
	} `yaml:"responder"`
	Cache struct {
		Interval *int `yaml:"interval"`
		Delay    *int `yaml:"delay"`
	} `yaml:"cache"`
	DB struct {
		DynamoDB *struct {
			Region           string `yaml:"region"`
			TableName        string `yaml:"table_name"`
			CAGsi            string `yaml:"ca_gsi"`
			Endpoint         string `yaml:"endpoint"`
			RetryMaxAttempts *int   `yaml:"retry_max_attempts"`
			Timeout          *int   `yaml:"timeout"`
		} `yaml:"dynamodb"`
		FileDB *struct {
			File string `yaml:"file"`
		} `yaml:"file"`
	} `yaml:"db"`
	HTTP struct {
		Port               string `yaml:"port"`
		Domain             string `yaml:"domain"`
		ReadTimeout        *int   `yaml:"read_timeout"`
		WriteTimeout       *int   `yaml:"write_timeout"`
		ReadHeaderTimeout  *int   `yaml:"read_header_timeout"`
		MaxHeaderBytes     *int   `yaml:"max_header_bytes"`
		MaxRequestBytes    *int   `yaml:"max_request_bytes"`
		CacheControlMaxAge *int   `yaml:"cache_control_max_age"`
	} `yaml:"http"`
}

// Supported CA DB type.
type CADBType int

const (
	// File DB.
	FileDBType CADBType = iota
	// DynamoDB.
	DynamoDBType
)

// Supported log format.
type LogFormat int

const (
	// JSON format.
	JSONFormat LogFormat = iota
	// Pretty format (human readable).
	PrettyFormat
)

// Default values.
const (
	ReadTimeOutDefault       = 30
	WriteTimeOutDefault      = 0
	ReadHeaderTimeoutDefault = 10
	MaxHeaderBytesDefault    = 1048576
	MaxRequestBytesDefault   = 256
	DynamoDBTimeoutDefault   = 60
	IntervalDefault          = 60
	DelayDefault             = 5
	LogLevelDefault          = "info"
	LogFormtDefault          = "json"
	ExpirationDefault        = "ignore"
)

// MissingParameterError is used when configuration paramemter is missing.
type MissingParameterError struct {
	Param string
}

func (e MissingParameterError) Error() string {
	return fmt.Sprintf(
		"'%s' parameter is not set or contains an empty value.", e.Param,
	)
}

// InvalidParameterError is used when configuration paramemter is invalid.
type InvalidParameterError struct {
	param       string
	description string
}

func (e InvalidParameterError) Error() string {
	return fmt.Sprintf(
		"'%s' parameter is invalid: %s",
		e.param,
		e.description,
	)
}

const (
	errsCap2  = 2
	errsCap4  = 4
	errsCap8  = 8
	errsCap16 = 6
	errsCap32 = 32
)

func markMissRequiredStr(sp string, spName string, errs []error) (string, []error) {
	if sp == "" {
		errs = append(errs, MissingParameterError{spName})
	}
	return sp, errs
}

// VerifyResponderConfig verifies .Log.
func (y ConfigYAML) VerifyLogConfig(cfg DyOCSPConfig) (DyOCSPConfig, []error) {
	nCfg := cfg
	errs := make([]error, 0, errsCap2)

	if y.Log.Level == "" {
		nCfg.LogLevel = LogLevelDefault
	} else if matched, _ := regexp.MatchString(`\A(?:error|warn|info|debug)\z`, y.Log.Level); !matched {
		errs = append(errs, InvalidParameterError{"log.level", "[error|warn|info|debug]"})
	} else {
		nCfg.LogLevel = y.Log.Level
	}

	switch nCfg.LogLevel {
	case "debug":
		nCfg.ZerologLevel = zerolog.DebugLevel
	case "info":
		nCfg.ZerologLevel = zerolog.InfoLevel
	case "warn":
		nCfg.ZerologLevel = zerolog.WarnLevel
	case "error":
		nCfg.ZerologLevel = zerolog.ErrorLevel
	}

	if y.Log.Format == "" {
		nCfg.LogFormat = LogFormtDefault
	} else if matched, _ := regexp.MatchString(`\A(?:json|pretty)\z`, y.Log.Format); !matched {
		errs = append(errs, InvalidParameterError{"log.format", "[json|pretty]"})
	} else {
		nCfg.LogFormat = y.Log.Format
	}

	switch nCfg.LogFormat {
	case "json":
		nCfg.ZerologFormat = JSONFormat
	case "pretty":
		nCfg.ZerologFormat = PrettyFormat
	}

	if len(errs) > 0 {
		return cfg, errs
	}
	return nCfg, nil
}

// VerifyResponderConfig verifies .Responder.
func (y ConfigYAML) VerifyResponderConfig(cfg DyOCSPConfig) (DyOCSPConfig, []error) {
	nCfg := cfg
	errs := make([]error, 0, errsCap4)

	// Responder.CA              Required
	nCfg.CA, errs = markMissRequiredStr(y.Responder.CA, "responder.ca", errs)
	// Responder.Certificate     Required
	nCfg.Certificate, errs = markMissRequiredStr(y.Responder.Certificate, "responder.responder_certificate", errs)
	// Responder.Key             Required (file or envionment variable)
	nCfg.Key = y.Responder.Key
	// Responder.Issuer          Required
	nCfg.Issuer, errs = markMissRequiredStr(y.Responder.Issuer, "responder.issuer_certificate", errs)

	if len(errs) > 0 {
		return cfg, errs
	}
	return nCfg, nil
}

// VerifyCacheConfig verifies .Caches.
func (y ConfigYAML) VerifyCacheConfig(cfg DyOCSPConfig) (DyOCSPConfig, []error) {
	nCfg := cfg
	errs := make([]error, 0, errsCap4)

	switch {
	case y.Cache.Interval == nil:
		nCfg.Interval = IntervalDefault
	case *y.Cache.Interval <= 0:
		errs = append(errs, InvalidParameterError{"cache.interval", "the number of seconds must be > 0"})
	default:
		nCfg.Interval = *y.Cache.Interval
	}

	switch {
	case y.Cache.Delay == nil:
		nCfg.Delay = DelayDefault
	case *y.Cache.Delay < 0:
		errs = append(errs, InvalidParameterError{"cache.delay", "the number of seconds must be >= 0"})
	case *y.Cache.Delay > nCfg.Interval:
		errs = append(errs, InvalidParameterError{"cache.delay", "cache.delay must be <= cache.interval"})
	default:
		nCfg.Delay = *y.Cache.Delay
	}

	if len(errs) > 0 {
		return cfg, errs
	}
	return nCfg, nil
}

// VerifyDynamoDBConfig verifies .DB.DynamoDB.
func (y ConfigYAML) VerifyDynamoDBConfig(cfg DyOCSPConfig) (DyOCSPConfig, []error) {
	nCfg := cfg
	errs := make([]error, 0, errsCap8)

	// DB.DynamoDB.Region            Required
	nCfg.DynamoDBRegion, errs = markMissRequiredStr(y.DB.DynamoDB.Region, "db.dynamodb.region", errs)
	// DB.DynamoDB.TableName         Required
	nCfg.DynamoDBTableName, errs = markMissRequiredStr(y.DB.DynamoDB.TableName, "db.dynamodb.table_name", errs)
	// DB.DynamoDB.CAGsi             Required
	nCfg.DynamoDBCAGsi, errs = markMissRequiredStr(y.DB.DynamoDB.CAGsi, "db.dynamodb.ca_gsi", errs)

	// DB.DynamoDB.Endpoint          Optional (default: DynamoDB Cloud)
	if y.DB.DynamoDB.Endpoint != "" {
		if matched, _ := regexp.MatchString(`\Ahttps?://`, y.DB.DynamoDB.Endpoint); !matched {
			errs = append(errs, InvalidParameterError{
				"db.dynamodb.endpoint",
				"url must start from 'http://' or 'https://'",
			})
		}
	}
	nCfg.DynamoDBEndpoint = y.DB.DynamoDB.Endpoint

	// DB.DynamoDB.RetryMaxAttempts  Optional (default: 0)
	switch {
	case y.DB.DynamoDB.RetryMaxAttempts == nil:
		nCfg.DynamoDBRetryMaxAttempts = 0
	case *y.DB.DynamoDB.RetryMaxAttempts < 0:
		errs = append(errs, InvalidParameterError{
			"db.dynamodb.retry_max_attempts",
			"the number of retries must be >= 0",
		})
	default:
		nCfg.DynamoDBRetryMaxAttempts = *y.DB.DynamoDB.RetryMaxAttempts
	}

	// DB.DynamoDB.RetryMaxAttempts  Optional (default: 0)
	switch {
	case y.DB.DynamoDB.Timeout == nil:
		nCfg.DynamoDBTimeout = DynamoDBTimeoutDefault
	case *y.DB.DynamoDB.Timeout <= 0:
		errs = append(errs, InvalidParameterError{
			"db.dynamodb.timeout",
			"the number of seconds for timeout must be > 0",
		})
	default:
		nCfg.DynamoDBTimeout = *y.DB.DynamoDB.Timeout
	}

	if len(errs) > 0 {
		return cfg, errs
	}
	return nCfg, nil
}

// VerifyFileDBConfig verifies .DB.FileDB.
func (y ConfigYAML) VerifyFileDBConfig(cfg DyOCSPConfig) (DyOCSPConfig, []error) {
	nCfg := cfg
	errs := make([]error, 0, errsCap2)

	// .DB.FileDB.File
	nCfg.FileDBFile, errs = markMissRequiredStr(y.DB.FileDB.File, "db.file.file", errs)

	if len(errs) > 0 {
		return cfg, errs
	}
	return nCfg, nil
}

// VerifyDBConfig verifies .DB.
func (y ConfigYAML) VerifyDBConfig(cfg DyOCSPConfig) (DyOCSPConfig, []error) {
	nCfg := cfg
	errs := make([]error, 0, errsCap16)

	dupN := 0

	// .DB.FileDB
	if y.DB.FileDB != nil {
		nCfg, errs = y.VerifyFileDBConfig(nCfg)
		nCfg.DBType = FileDBType
		dupN++
	}

	// .DB.DynamoDB
	if y.DB.DynamoDB != nil {
		nCfg, errs = y.VerifyDynamoDBConfig(nCfg)
		nCfg.DBType = DynamoDBType
		dupN++
	}

	if dupN == 0 {
		errs = []error{MissingParameterError{"db.<db-type>"}}
		return cfg, errs
	}

	if dupN > 1 {
		errs = []error{InvalidParameterError{"db.<db-type>", "DB type is exclusive"}}
		return cfg, errs
	}

	if len(errs) > 0 {
		return cfg, errs
	}
	return nCfg, nil
}

func specOrDefInt(i *int, def int) int {
	if i == nil {
		return def
	}
	return *i
}

// VerifyHTTPConfig verifies .HTTP.
func (y ConfigYAML) VerifyHTTPConfig(cfg DyOCSPConfig) (DyOCSPConfig, []error) {
	nCfg := cfg
	errs := make([]error, 0, errsCap4)

	// HTTP.Port                 Optional
	if y.HTTP.Port == "" {
		nCfg.Port = "80"
	} else if matched, _ := regexp.MatchString(`\A[1-9][0-9]*\z`, y.HTTP.Port); !matched {
		errs = append(errs, InvalidParameterError{"http.port", "must be the valid port number"})
	} else {
		nCfg.Port = y.HTTP.Port
	}
	// HTTP.Domain               Optional
	nCfg.Domain = y.HTTP.Domain
	// HTTP.ReadTimeout          Optional
	nCfg.ReadTimeout = specOrDefInt(y.HTTP.ReadTimeout, ReadTimeOutDefault)
	// HTTP.WriteTimeout         Optional
	nCfg.WriteTimeout = specOrDefInt(y.HTTP.WriteTimeout, WriteTimeOutDefault)
	// HTTP.ReadHeaderTimeout    Optional
	nCfg.ReadHeaderTimeout = specOrDefInt(y.HTTP.ReadHeaderTimeout, ReadHeaderTimeoutDefault)
	// HTTP.MaxHeaderBytes       Optional
	nCfg.MaxHeaderBytes = specOrDefInt(y.HTTP.MaxHeaderBytes, MaxHeaderBytesDefault)
	// HTTP.MaxHeaderBytes       Optional
	nCfg.MaxRequestBytes = specOrDefInt(y.HTTP.MaxRequestBytes, MaxRequestBytesDefault)

	switch {
	case y.HTTP.CacheControlMaxAge == nil:
		nCfg.CacheControlMaxAge = nCfg.Interval
	case *y.HTTP.CacheControlMaxAge <= 0:
		errs = append(
			errs,
			InvalidParameterError{"http.cache_control_max_age", "cache-control max-age must be > 0"},
		)
	case *y.HTTP.CacheControlMaxAge > nCfg.Interval:
		errs = append(
			errs,
			InvalidParameterError{"http.cache_control_max_age", "cache-control max-age must be <= cache.interval"},
		)
	default:
		nCfg.CacheControlMaxAge = *y.HTTP.CacheControlMaxAge
	}

	if len(errs) > 0 {
		return cfg, errs
	}
	return nCfg, nil
}

// Verify verifies the configuration root '.' using the ConfigYAML.Verify* methods.
// If it detects any invalid parameters, it returns an error slice.
// If there are no errors, it returns nil.
func (y ConfigYAML) Verify(cfg DyOCSPConfig) (DyOCSPConfig, []error) {
	nCfg := cfg
	errs := make([]error, 0, errsCap32)

	// .Version  Required 0.1 only
	nCfg.Version, errs = markMissRequiredStr(y.Version, "version", errs)
	if y.Version != "" {
		if matched, _ := regexp.MatchString(`\A(?:0.1)\z`, y.Version); !matched {
			errs = append(errs, InvalidParameterError{"version", "[0.1]"})
		}
	}

	// .Strict  Optional
	nCfg.Strict = y.Strict

	// .Expiration  Optional
	if y.Expiration == "" {
		nCfg.Expiration = ExpirationDefault
	} else if matched, _ := regexp.MatchString(`\A(?:ignore|warn|invalid)\z`, y.Expiration); !matched {
		errs = append(errs, InvalidParameterError{"expiration", "[ignore|warn|invalid]"})
	} else {
		nCfg.Expiration = y.Expiration
	}

	// .Log.Level  Optional error, warn, info, debug (default: info)
	nCfg, logErrs := y.VerifyLogConfig(nCfg)
	if logErrs != nil {
		errs = append(errs, logErrs...)
	}

	// .Responder
	nCfg, responderErrs := y.VerifyResponderConfig(nCfg)
	if responderErrs != nil {
		errs = append(errs, responderErrs...)
	}

	// .Cache
	nCfg, cacheErrs := y.VerifyCacheConfig(nCfg)
	if cacheErrs != nil {
		errs = append(errs, cacheErrs...)
	}

	// .DB
	nCfg, dbErrs := y.VerifyDBConfig(nCfg)
	if dbErrs != nil {
		errs = append(errs, dbErrs...)
	}

	// .HTTP
	nCfg, httpErrs := y.VerifyHTTPConfig(nCfg)
	if httpErrs != nil {
		errs = append(errs, httpErrs...)
	}

	if len(errs) > 0 {
		return cfg, errs
	}
	return nCfg, nil
}
